### DISCUSSION TITLE
feat: API of t.procedure.query/mutation

### DISCUSSION THREAD

**Discussion by olehmisar** [trpc/trpc#2940] [Q&A] [answered]
Why does the API look like this:
```ts
t.procedure
  .input(...)
  .query(...)
```
?

Did you consider this API:
```ts
t.query({
  input: z.object(...),
  async resolve() { ... },
}) 
```

It seems more natural to me and looks like v9
### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

**Comment by KATT** [comment, marked as answer]
With this structure you can define if you want to have the inputs before or after a middlewares & you can also mix it with having multiple input parsers and making many base procedures that look similar

**Reply by KATT** [reply]
I am adding some sections on this in the docs, see the below links for some ideas that it enables:

- https://trpc.io/docs/v10/procedures#reusable-base-procedures
- https://trpc.io/docs/v10/procedures#multiple-input-parsers
- https://trpc.io/docs/v10/metadata#example-with-per-route-authentication-settings

**Reply by olehmisar** [reply]
first two links do not work

**Reply by KATT** [reply]
Updated
